src/app-logic/url-handling.js

```diff
@@ -462,11 +462,13 @@ export function urlFromState(urlState: UrlState): string {
 }
 
 export function getDataSourceFromPathParts(pathParts: string[]): DataSource {
-  const str = pathParts[0] || 'none';
-  // With this switch, flow is able to understand that we return a valid value
+  // By casting `str` to a DataSource beforehand, we let Flow enforce that we
+  // look at all possible values.
+  const str: DataSource = (pathParts[0]: any) || 'none';
```

Comment:
This is a small change from before. I got fed up that this `switch` wasn't checked by Flow. Unintuitively adding this `any` will make this possible.

Reply:
Suggestion: This makes a little nervous. I have two possible suggestions here:

### Suggestion A

A more complete solution would be to create a `toValidDataSource()` function, see `toValidTabSlug()` for an example. Then the different cases can be handled explicitly. This would be more verbose.

or

### Suggestion B

A looser, but more type-safe option would be to use the coerce utility:

`coerce<string | void, DataSource>(pathParts[0])`

Reply:
about suggestion A => Yes, that's the goal of this function actually, see below. But because the input is a string, Flow doesn't warn here if we don't add a `case` when we have a new datasource. This change is so that Flow actually warns about it.

About suggestion B: good idea!

Reply:
Actually `toValidTabSlug` uses the same trick, just a bit differently:
```js
export function toValidTabSlug(tabSlug: any): TabSlug | null {
  const coercedTabSlug = (tabSlug: TabSlug);
```

Reply:
I changed the name of the function to something that's closer to our current practices, and used `coerce`.